src/hotspot/cpu/x86/universalNativeInvoker_x86.cpp

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "asm/macroAssembler.hpp"
+#include "code/codeBlob.hpp"
+#include "memory/resourceArea.hpp"
+#include "prims/universalNativeInvoker.hpp"
+
+void ProgrammableInvoker::Generator::generate() {
+  __ enter();
+
+  // Put the context pointer in ebx/rbx - it's going to be heavily used below both before and after the call
+  Register ctxt_reg = rbx;
+  Register used_regs[] = { ctxt_reg, rcx, rsi, rdi };
+  GrowableArray<Register> preserved_regs;
+
+  for (size_t i = 0; i < sizeof(used_regs)/sizeof(Register); i++) {
+    Register used_reg = used_regs[i];
+    if (!_abi->is_volatile_reg(used_reg)) {
+      preserved_regs.push(used_reg);
+    }
+  }
+
+  __ block_comment("init_and_alloc_stack");
+
+  for (int i = 0; i < preserved_regs.length(); i++) {
+    __ push(preserved_regs.at(i));
+  }
+
+  __ movptr(ctxt_reg, c_rarg0); // FIXME c args? or java?
+
+  __ block_comment("allocate_stack");
+  __ movptr(rcx, Address(ctxt_reg, (int) _layout->stack_args_bytes));
+  __ subptr(rsp, rcx);
+  __ andptr(rsp, -_abi->_stack_alignment_bytes);
+
+  // Note: rcx is used below!
+
+
+  __ block_comment("load_arguments");
+
+  __ shrptr(rcx, LogBytesPerWord); // bytes -> words
+  __ movptr(rsi, Address(ctxt_reg, (int) _layout->stack_args));
+  __ movptr(rdi, rsp);
+  __ rep_mov();
+
+
+  for (int i = 0; i < _abi->_vector_argument_registers.length(); i++) {
+    // [1] -> 64 bit -> xmm
+    // [2] -> 128 bit -> xmm
+    // [4] -> 256 bit -> ymm
+    // [8] -> 512 bit -> zmm
+
+    XMMRegister reg = _abi->_vector_argument_registers.at(i);
+    size_t offs = _layout->arguments_vector + i * sizeof(VectorRegister);
+    if (UseAVX >= 3) {
```

Comment:
Assuming worst-case scenario (w.r.t. value size) is sub-optimal. Considering FP values are handled as vectors you end up operating on 32-bit/64-bit values as if they were 512-bit in size. And, in addition to wasted memory bandwidth, EVEX-encoded instructions may trigger CPU frequency scaling which will penalize cases when AVX512 is not used. So, it is worth considering annotating vector values with their actual sizes and taking the size into account when operating on vectors.

Reply:
Yes, more cleanup is needed here. We don't support vectors at all right now, so I'd rather remove this code and only operate on XMM registers instead.

In the future this could be handled using separate VMStorage types for different vector sizes.